### PR TITLE
Adds support for optional kwargs in the EKS Operators

### LINKED ISSUE #19641: EKSCreateNodegroupOperator - missing kwargs
### Description

Boto3  / eks / create_nodegroup api supports the following kwargs:

```python
   clusterName='string',
    nodegroupName='string',
    scalingConfig={
        'minSize': 123,
        'maxSize': 123,
        'desiredSize': 123
    },
    diskSize=123,
    subnets=[
        'string',
    ],
    instanceTypes=[
        'string',
    ],
    amiType='AL2_x86_64'|'AL2_x86_64_GPU'|'AL2_ARM_64'|'CUSTOM'|'BOTTLEROCKET_ARM_64'|'BOTTLEROCKET_x86_64',
    remoteAccess={
        'ec2SshKey': 'string',
        'sourceSecurityGroups': [
            'string',
        ]
    },
    nodeRole='string',
    labels={
        'string': 'string'
    },
    taints=[
        {
            'key': 'string',
            'value': 'string',
            'effect': 'NO_SCHEDULE'|'NO_EXECUTE'|'PREFER_NO_SCHEDULE'
        },
    ],
    tags={
        'string': 'string'
    },
    clientRequestToken='string',
    launchTemplate={
        'name': 'string',
        'version': 'string',
        'id': 'string'
    },
    updateConfig={
        'maxUnavailable': 123,
        'maxUnavailablePercentage': 123
    },
    capacityType='ON_DEMAND'|'SPOT',
    version='string',
    releaseVersion='string'
```

The current implementation of the operator support the following kwargs only:

```python
   clusterName='string',
   nodegroupName='string',
   subnets=[ 'string', ],
   nodeRole='string',
```
   

### Use case/motivation

With the current implementation of the Operator you can bring up basic EKS nodegroup which is not useful. I want to fully configure my nodegroup with this operation like I can do it with boto3.

### Related issues

Continue of #19372

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by ferruzzi**
closes: https://github.com/apache/airflow/issues/19641

TL;DR:   BaseOperator will throw an exception if it gets an unexpected parameter so this is used as a workaround.

Adds the option to include an operator-specific kwargs parameter to the three "Create" operators in the EKS module as seen in other Operators.  Includes unit testing and some formatting adjustments to related files to bring them in line with formatting in other modules.

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

### CODE REVIEW

File: airflow/providers/amazon/aws/hooks/eks.py [hunk 1]
```diff
@@ -135,13 +146,12 @@ def create_nodegroup(
...
+
         # The below tag is mandatory and must have a value of either 'owned' or 'shared'
         # A value of 'owned' denotes that the subnets are exclusive to the nodegroup.
         # The 'shared' value allows more than one resource to use the subnet.
-        tags = {'kubernetes.io/cluster/' + clusterName: 'owned'}
-        if "tags" in kwargs:
-            tags = {**tags, **kwargs["tags"]}
-            kwargs.pop("tags")
+        tags = kwargs.pop("tags", {})
+        tags[f'kubernetes.io/cluster/{clusterName}'] = 'owned'
```
Comment (dstandish) [inline airflow/providers/amazon/aws/hooks/eks.py]: this is breaking because now user cannot set it to 'shared'

File: airflow/providers/amazon/aws/operators/eks.py [hunk 2]
```diff
@@ -89,36 +91,45 @@ class EksCreateClusterOperator(BaseOperator):
          If this is None or empty then the default boto3 behaviour is used.
     :type region: str
 
-    If compute is assigned the value of 'nodegroup', the following are required:
+    If compute is assigned the value of 'nodegroup':
```
Comment (dstandish) [inline airflow/providers/amazon/aws/operators/eks.py:94]: why this change?  seems better before, more explicit about the meaning.

if x, then _what_? rather than just "if x:"

perhaps it was not correct before?

File: airflow/providers/amazon/aws/hooks/eks.py (hunk 1 above)
Comment (uranusjr) [inline airflow/providers/amazon/aws/hooks/eks.py]: How about changing this function to

```python
def create_nodegroup(
    self,
    clusterName: str,
    nodegroupName: str,
    subnets: List[str],
    nodeRole: str,
    *,
    tags: Optional[dict] = None
    **kwargs,
) -> Dict:
    ...
    if tags is not None:
        resolved_tags = tags
    else:
        resolved_tags = {}
    resolved_tags[f'kubernetes.io/cluster/{clusterName}'] = 'owned'
```

to better leverage Python’s built-in syntax. The `kwargs.pop()` or `if key in kwargs` pattern was a pragmatic workaround on Python 2, but now that keyword-only arguments are available, they make the API both easier to read and write.

Comment (ferruzzi) [reply airflow/providers/amazon/aws/hooks/eks.py]: Thanks, I didn't think of that syntax for the signature.   If we're going that route, how would you feel about simplifying it to 

```
resolved_tags = tags or {}
resolved_tags[f'kubernetes.io/cluster/{clusterName}'] = 'owned'
```

File: airflow/providers/amazon/aws/operators/eks.py (hunk 2 above)
Comment (ferruzzi) [reply airflow/providers/amazon/aws/operators/eks.py:94]: We've now added some that are available but non-required.  The ones that were there are required if you define compute='nodegroup' but `create_nodegroup_kwargs` is optional with that set.  Can you suggest a concise way to rephrase it that still makes sense?  The ideas I came up with were long and/or awkward, and this sounded better.

File: airflow/providers/amazon/aws/hooks/eks.py (hunk 1 above)
Comment (ferruzzi) [reply airflow/providers/amazon/aws/hooks/eks.py]: I thought of that when I was looking at @uranusjr 's comment above but wanted to test it to see.  Would it have preserved the user's setting the way it was before?

Comment (uranusjr) [reply airflow/providers/amazon/aws/hooks/eks.py]: I don’t generally like the `x or default` pattern, but this is OK as well.

Comment (uranusjr) [reply airflow/providers/amazon/aws/hooks/eks.py]: If “before” means the code currently in `main`, yes it would, because `{**tags, **kwargs["tags"]}` means that if any key in `kwargs["tags"]` would override the value under the same key in `tags`.

```python
>>> a = {"x": 1}
>>> b = {"x": 2}
>>> {**a, **b}
{"x": 2}
```

So the backward compatible implementation would be something like

```python
resolved_tags = tags or {}
cluster_tag_key = f'kubernetes.io/cluster/{clusterName}'
if cluster_tag_key not in resolved_tags:
    resolved_tags[cluster_tag_key] = 'owned'
```

or `setdefault` might work as well.

Comment (dstandish) [reply airflow/providers/amazon/aws/hooks/eks.py]: in effect the current code has a default of `owned` and i think that should be preserved without good reason for changing

Comment (ferruzzi) [reply airflow/providers/amazon/aws/hooks/eks.py]: Yes, I couldn't remember offhand if the previous/existing implementation `{**tags, **kwargs["tags"]}` overwrote the default with the user's or not.  I was going to confirm how it worked before I said anything, but @dstandish caught it first.

I'll get this and the above corrected in the morning.

File: airflow/providers/amazon/aws/operators/eks.py [hunk 3]
```diff
@@ -89,36 +91,45 @@ class EksCreateClusterOperator(BaseOperator):
...
+    If compute is assigned the value of 'fargate':
+
+    :param fargate_profile_name: *REQUIRED* The unique name to give your AWS Fargate profile. (templated)
     :type fargate_profile_name: str
-    :param fargate_pod_execution_role_arn: The Amazon Resource Name (ARN) of the pod execution role to
-         use for pods that match the selectors in the AWS Fargate profile. (templated)
+    :param fargate_pod_execution_role_arn: *REQUIRED* The Amazon Resource Name (ARN) of the pod execution
+         role to use for pods that match the selectors in the AWS Fargate profile. (templated)
     :type podExecutionRoleArn: str
     :param selectors: The selectors to match for pods to use this AWS Fargate profile. (templated)
```
Comment (dstandish) [inline airflow/providers/amazon/aws/operators/eks.py]: unrelated but i think this is supposed to be `fargate_selectors`

File: airflow/providers/amazon/aws/operators/eks.py (hunk 2 above)
Comment (dstandish) [reply airflow/providers/amazon/aws/operators/eks.py:94]: yeah i understand -- some are optional now

fine with me

File: airflow/providers/amazon/aws/operators/eks.py [hunk 4]
```diff
@@ -89,36 +91,45 @@ class EksCreateClusterOperator(BaseOperator):
...
         "resources_vpc_config",
+        "create_cluster_kwargs",
         "compute",
         "nodegroup_name",
         "nodegroup_role_arn",
+        "create_nodegroup_kwargs",
         "fargate_profile_name",
         "fargate_pod_execution_role_arn",
         "fargate_selectors",
+        "create_fargate_profile_kwargs",
```
Comment (dstandish) [inline airflow/providers/amazon/aws/operators/eks.py:132]: what do you think about dropping the `create_` so it would juts be `cluster_kwargs`, `nodegroup_kwargs`, `fargate_profile_kwargs`?

it's a bit less wordy, but i think still clear enough given that this a "create cluster" operator

i'm also tempted to suggest deprecating non-kwargs params (so in future we'd can only use the kwargs params) but maybe people usually only need the existing params and it's convenient to have them?

Comment (ferruzzi) [reply airflow/providers/amazon/aws/operators/eks.py:132]: > what do you think about dropping the create_ so it would juts be cluster_kwargs, nodegroup_kwargs, fargate_profile_kwargs?

Could do that.  I left it long like this so we didn't paint ourselves into a corner.  I was thinking this way gave flexibility in case other endpoints got kwargs later, or maybe we go back and decide to paginate the list operators and need a list_clusters_kwargs or something, but that's maybe needless future-proofing?  

> i'm also tempted to suggest deprecating non-kwargs params (so in future we'd can only use the kwargs params) but maybe people usually only need the existing params and it's convenient to have them?

Yeah, that was something I discussed with @o-nikolas.  That or implementing a hybrid where you could use them explicitly or within the cluster_kwargs and the logic would handle it for you either way, but that made the code significantly (and needlessly?) more complicated so we settled on this as the easiest non-breaking answer.   What do you think?

File: airflow/providers/amazon/aws/operators/eks.py (hunk 3 above)
Comment (ferruzzi) [reply airflow/providers/amazon/aws/operators/eks.py]: Indeed.  I'll get that updated while I'm in there.

File: airflow/providers/amazon/aws/hooks/eks.py (hunk 1 above)
Comment (ferruzzi) [reply airflow/providers/amazon/aws/hooks/eks.py]: Resolving this one since we picked up the conversation below in dstandish's thread.

File: airflow/providers/amazon/aws/operators/eks.py (hunk 4 above)
Comment (dstandish) [reply airflow/providers/amazon/aws/operators/eks.py:132]: > Could do that. I left it long like this so we didn't paint ourselves into a corner. I was thinking this way gave flexibility in case other endpoints got kwargs later, or maybe we go back and decide to paginate the list operators and need a list_clusters_kwargs or something, but that's maybe needless future-proofing?

i leave it up to you.  i like the shorter form better but being verbose is fine too.

> Yeah, that was something I discussed with @o-nikolas. That or implementing a hybrid where you could use them explicitly or within the cluster_kwargs and the logic would handle it for you either way, but that made the code significantly (and needlessly?) more complicated so we settled on this as the easiest non-breaking answer. What do you think?

yeah if we were to go this route we'd have to go hybrid first (and emit a warning whenever user supplied the non-kwargs params).  and then in next major release of the provider we would chop the params.  if you want to proceed down that route we could deal with that in a later PR does not have to be dealt with here.

File: airflow/providers/amazon/aws/hooks/eks.py (hunk 1 above)
Comment (ferruzzi) [reply airflow/providers/amazon/aws/hooks/eks.py]: @uranusjr - Is there a particular reason why you are making the tags parameter keyword-only in your solution here?

**Comment by ferruzzi** [comment]
[42751d7](https://github.com/apache/airflow/pull/20819/commits/6d059d4af4fcaeae0d3980b49d86c73cf42751d7) should address the tag override issue, and [a84ad15](https://github.com/apache/airflow/pull/20819/commits/a84ad15e87e82b240440991ba060956ecda434d0) adds two unit tests to make sure it doesn't happen again.

File: airflow/providers/amazon/aws/operators/eks.py (hunk 3 above)
Comment (ferruzzi) [reply airflow/providers/amazon/aws/operators/eks.py]: Resolved in https://github.com/apache/airflow/pull/20819/commits/6d059d4af4fcaeae0d3980b49d86c73cf42751d7

File: airflow/providers/amazon/aws/operators/eks.py (hunk 4 above)
Comment (ferruzzi) [reply airflow/providers/amazon/aws/operators/eks.py:132]: Let's leave deprecating that for the future.  I'd like to come up with some consistent plan on how we're going to handle these throughout the AWS provider package (and beyond?) before we start making breaking changes to one service.

File: airflow/providers/amazon/aws/hooks/eks.py (hunk 1 above)
Comment (uranusjr) [reply airflow/providers/amazon/aws/hooks/eks.py]: The parameter was retrieved with `kwargs.pop()`, so it can only be passed as a keyword argument to begin with. I don’t have particular reasons to make it keyword-only, but since it already is _logically_ keyword-only, the keyword-only syntax makes the intention more explicit.

Comment (ferruzzi) [reply airflow/providers/amazon/aws/hooks/eks.py]: I guess that makes sense.  I've added that in https://github.com/apache/airflow/pull/20819/commits/cc12d26eb4fea209ae5c616feaffa94c95cf8a62, just wanted to make sure I understood why.  Thanks

**Comment by ferruzzi** [comment]
AFAIK, all concerns raised so far have been addressed at this point.

**Review by uranusjr — APPROVED** [review]

**Review by kaxil — APPROVED** [review]
